browser/brave_vpn/win/wireguard_utils_win.cc

```diff
@@ -22,16 +22,84 @@
 #include "base/task/single_thread_task_runner.h"
 #include "base/task/thread_pool.h"
 #include "base/win/com_init_util.h"
+#include "base/win/registry.h"
 #include "base/win/scoped_bstr.h"
 #include "brave/browser/brave_vpn/win/service_constants.h"
 #include "brave/browser/brave_vpn/win/service_details.h"
 #include "brave/components/brave_vpn/common/win/utils.h"
 #include "brave/components/brave_vpn/common/wireguard/win/brave_wireguard_manager_idl.h"
+#include "url/gurl.h"
 
 namespace brave_vpn {
 
 namespace {
 std::optional<bool> g_wireguard_service_registered_for_testing;
+// TODO(bsclifton): consider persisting last used value to registry or disk.
+// As-is, it won't survive the owning process being restarted.
+std::optional<std::string> g_smart_proxy_url;
+constexpr wchar_t kSystemProxyRegistryKey[] =
+    L"Software\\Microsoft\\Windows\\CurrentVersion\\Internet Settings";
+
+void MaybeEnableSystemProxy() {
+  if (!g_smart_proxy_url.has_value()) {
+    return;
+  }
+
+  base::win::RegKey key;
+  LONG rv =
+      key.Create(HKEY_CURRENT_USER, kSystemProxyRegistryKey, KEY_ALL_ACCESS);
+  if (rv == ERROR_SUCCESS && key.Valid()) {
+    std::wstring current_proxy_url;
+    if (key.ReadValue(L"AutoConfigURL", &current_proxy_url) == ERROR_SUCCESS) {
+      // Person already has a proxy value set; don't touch it.
+      VLOG(1) << L"User has `AutoConfigURL` value already set: \""
+              << current_proxy_url
+              << "\".\nBrave VPN smart proxy routing can't be used.";
+    } else {
+      // Set the system proxy URL.
+      current_proxy_url = base::UTF8ToWide(*g_smart_proxy_url);
+      rv = key.WriteValue(L"AutoConfigURL", current_proxy_url.c_str());
+      VLOG(1) << L"Set Brave VPN smart proxy routing to = \""
+              << current_proxy_url << "\"";
+    }
+  }
+}
+
+void MaybeDisableSystemProxy() {
+  base::win::RegKey key;
+  LONG rv =
+      key.Create(HKEY_CURRENT_USER, kSystemProxyRegistryKey, KEY_ALL_ACCESS);
+  if (rv != ERROR_SUCCESS || !key.Valid()) {
+    VLOG(1) << L"Nothing to clean up; user does not have a key for "
+               L"`AutoConfigURL`.";
+    return;
+  }
+
+  // Only disable system proxy if...
+  // 1) There's a value actually set.
+  std::wstring current_proxy_url;
+  if (key.ReadValue(L"AutoConfigURL", &current_proxy_url) != ERROR_SUCCESS) {
+    VLOG(1) << L"Error reading `AutoConfigURL`; value may not exist.";
+    return;
+  }
+
+  // 2) The value is a valid URL.
+  GURL current_proxy_gurl(base::WideToUTF8(current_proxy_url));
+  if (current_proxy_gurl.is_empty() || !current_proxy_gurl.is_valid()) {
```

Comment:
I think checking `is_valid()` is same with above as `is_valid()` gives false if it's empty. 
```suggestion
  if (!current_proxy_gurl.is_valid()) {
```

Reply:
Fixed!